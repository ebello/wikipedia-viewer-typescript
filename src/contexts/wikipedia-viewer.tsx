import React, { useState, useEffect, ReactNode } from 'react';
import PropTypes from 'prop-types';

interface ViewingHistoryItem {
  timestamp: number,
  title: string,
}

interface WikipediaContextValue {
  readonly viewingHistory: ViewingHistoryItem[],
  addToViewingHistory: (title: string) => void,
  clearViewingHistory: () => void,
}

type ContextProps = {
  children: ReactNode,
};

const WikipediaViewerContext = React.createContext<WikipediaContextValue>({
  viewingHistory: [],
  addToViewingHistory: () => null,
  clearViewingHistory: () => null,
});

const WikipediaViewer = ({ children }: ContextProps): JSX.Element => {
  const viewingHistoryStr = window?.localStorage.getItem('viewingHistory');
  const defaultHistory = JSON.parse(viewingHistoryStr || '') as ViewingHistoryItem[] || [];
  const [viewingHistory, setViewingHistory] = useState(defaultHistory);

  const addToViewingHistory = (title: string) => {
    setViewingHistory([...viewingHistory, {
      timestamp: Date.now(),
      title,
    }]);
  };

  const clearViewingHistory = () => setViewingHistory([]);

  useEffect(() => {
    window.localStorage.setItem('viewingHistory', JSON.stringify(viewingHistory));
  }, [viewingHistory]);

  return (
    <WikipediaViewerContext.Provider
      value={{
        viewingHistory,
        addToViewingHistory,
        clearViewingHistory,
      }}
    >
      {children}
    </WikipediaViewerContext.Provider>
  );
};

WikipediaViewer.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default WikipediaViewer;

export { WikipediaViewerContext };
