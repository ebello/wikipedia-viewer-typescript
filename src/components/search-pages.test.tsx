import React from 'react';
import { RenderResult, waitFor } from '@testing-library/react';
import { History } from '@reach/router';
import {
  renderWithRouter, screen, fireEvent, sleep,
} from '../test-utils';
import { WikipediaViewerContext } from '../contexts/wikipedia-viewer';
import SearchPages from './search-pages';

describe('Search Pages', () => {
  it('renders with no results or history', () => {
    const { baseElement } = renderWithRouter(<SearchPages />);
    fireEvent.focus(screen.getByRole('combobox'));
    expect(baseElement).toMatchSnapshot();
  });

  describe('with viewingHistory', () => {
    let renderResult: RenderResult & { history: History };
    beforeEach(() => {
      renderResult = renderWithRouter(
        <WikipediaViewerContext.Provider
          value={{
            viewingHistory: [
              { timestamp: 1234, title: 'sample title 2' },
              { timestamp: 234, title: 'sample title' },
              { timestamp: 12345, title: 'sample title' },
              { timestamp: 88888, title: 'other' },
            ],
            addToViewingHistory: () => {},
            clearViewingHistory: () => {},
          }}
        >
          <SearchPages />
        </WikipediaViewerContext.Provider>,
      );
    });

    it('renders', () => {
      const { baseElement } = renderResult;
      fireEvent.focus(screen.getByRole('combobox'));
      expect(baseElement).toMatchSnapshot();
    });

    it('only shows unique titles', () => {
      fireEvent.focus(screen.getByRole('combobox'));
      expect(screen.getAllByRole('option')).toHaveLength(3);
    });

    it('sorts by most recent first', () => {
      fireEvent.focus(screen.getByRole('combobox'));
      const options = screen.getAllByRole('option');
      expect(options[0]).toHaveTextContent('other');
      expect(options[1]).toHaveTextContent('sample title');
      expect(options[2]).toHaveTextContent('sample title 2');
    });

    it('filters by search query', async () => {
      fireEvent.change(screen.getByRole('combobox'), { target: { value: 'ot' } });
      await waitFor(() => {
        const options = screen.getAllByRole('option');
        expect(options).toHaveLength(1);
        expect(options[0]).toHaveTextContent('other');
      });
    });

    it('navigates to the page', () => {
      fireEvent.click(screen.getByText('other'));
      const { history } = renderResult;
      expect(history.location.pathname).toBe('/wiki/other');
    });
  });

  describe('with search results', () => {
    let renderResult: RenderResult & { history: History };
    beforeEach(() => {
      renderResult = renderWithRouter(<SearchPages />);
      fireEvent.change(screen.getByRole('combobox'), { target: { value: 'test' } });
    });

    it('renders', async () => {
      await waitFor(() => {
        const options = screen.getAllByRole('option');
        expect(options).toHaveLength(2);
        expect(options[0]).toHaveTextContent('test');
        expect(options[1]).toHaveTextContent('testing');
        const { baseElement } = renderResult;
        expect(baseElement).toMatchSnapshot();
      });
    });

    it('navigates to the page', async () => {
      await waitFor(() => {
        const options = screen.getAllByRole('option');
        fireEvent.click(options[1]);
        const { history } = renderResult;
        expect(history.location.pathname).toBe('/wiki/testing');
      });
    });

    it('does not set search results if unmounted', async () => {
      const { unmount } = renderResult;
      unmount();
      await sleep(1000);
      expect(screen.queryAllByRole('option')).toHaveLength(0);
    });
  });
});
