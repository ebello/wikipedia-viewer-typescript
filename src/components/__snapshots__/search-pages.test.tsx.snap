// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Pages renders with no results or history 1`] = `
.emotion-1 {
  display: block;
  width: 100%;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    display: inline-block;
    width: 50%;
  }
}

.emotion-0 {
  width: 100%;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: 24px;
  }
}

<body>
  <div>
    <div
      class="emotion-1"
      data-reach-combobox=""
    >
      <input
        aria-autocomplete="both"
        aria-controls="listbox--1"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-label="Pages"
        class="emotion-0"
        data-reach-combobox-input=""
        placeholder="Search Wikipedia"
        role="combobox"
        value=""
      />
    </div>
  </div>
</body>
`;

exports[`Search Pages with search results renders 1`] = `
.emotion-1 {
  display: block;
  width: 100%;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    display: inline-block;
    width: 50%;
  }
}

.emotion-0 {
  width: 100%;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: 24px;
  }
}

.emotion-2 {
  margin-top: 4px;
  margin-bottom: 4px;
  margin-left: 8px;
  color: primary;
}

<body>
  <div>
    <div
      class="emotion-1"
      data-reach-combobox=""
    >
      <input
        aria-autocomplete="both"
        aria-controls="listbox--36"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-label="Pages"
        class="emotion-0"
        data-reach-combobox-input=""
        placeholder="Search Wikipedia"
        role="combobox"
        value="test"
      />
    </div>
  </div>
  <reach-portal>
    <div
      data-reach-combobox-popover=""
      data-reach-popover=""
      style="position: absolute; width: 0px; left: 0px; top: 0px;"
      tabindex="-1"
    >
      <ul
        data-reach-combobox-list=""
        id="listbox--36"
        role="listbox"
      >
        <h5
          class="emotion-2"
        >
          Search results
        </h5>
        <li
          aria-selected="false"
          data-reach-combobox-option=""
          id="3556498"
          role="option"
          tabindex="-1"
        >
          <span
            data-user-value="true"
          >
            test
          </span>
        </li>
        <li
          aria-selected="false"
          data-reach-combobox-option=""
          id="-1422446064"
          role="option"
          tabindex="-1"
        >
          <span
            data-user-value="true"
          >
            test
          </span>
          <span
            data-suggested-value="true"
          >
            ing
          </span>
        </li>
      </ul>
    </div>
  </reach-portal>
</body>
`;

exports[`Search Pages with viewingHistory renders 1`] = `
.emotion-1 {
  display: block;
  width: 100%;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    display: inline-block;
    width: 50%;
  }
}

.emotion-0 {
  width: 100%;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: 24px;
  }
}

.emotion-3 {
  margin-top: 8px;
  margin-bottom: 8px;
  margin-left: 8px;
  color: primary;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: inline-block;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
  background-color: primary;
  border: 0;
  border-radius: 4px;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 8px;
  padding-right: 8px;
  font-size: 85%;
  margin-left: 8px;
  background-color: bgmuted;
}

<body>
  <div>
    <div
      class="emotion-1"
      data-reach-combobox=""
    >
      <input
        aria-autocomplete="both"
        aria-controls="listbox--4"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-label="Pages"
        class="emotion-0"
        data-reach-combobox-input=""
        placeholder="Search Wikipedia"
        role="combobox"
        value=""
      />
    </div>
  </div>
  <reach-portal>
    <div
      data-reach-combobox-popover=""
      data-reach-popover=""
      style="position: absolute; width: 0px; left: 0px; top: 0px;"
      tabindex="-1"
    >
      <ul
        data-reach-combobox-list=""
        id="listbox--4"
        role="listbox"
      >
        <h5
          class="emotion-3"
        >
          Recent history
          <button
            class="emotion-2"
          >
            Clear history
          </button>
        </h5>
        <li
          aria-selected="false"
          data-reach-combobox-option=""
          id="106069776"
          role="option"
          tabindex="-1"
        >
          <span
            data-suggested-value="true"
          >
            other
          </span>
        </li>
        <li
          aria-selected="false"
          data-reach-combobox-option=""
          id="-1347533726"
          role="option"
          tabindex="-1"
        >
          <span
            data-suggested-value="true"
          >
            sample title
          </span>
        </li>
        <li
          aria-selected="false"
          data-reach-combobox-option=""
          id="2100213748"
          role="option"
          tabindex="-1"
        >
          <span
            data-suggested-value="true"
          >
            sample title 2
          </span>
        </li>
      </ul>
    </div>
  </reach-portal>
</body>
`;
